### PR TITLE
Make reports respect timeouts

### PR DESCRIPTION
**Pull request by speezepearson**
Not too complicated, just long! Had to plumb a new field all the way from UI to database.

Changes, sorted by how far down you'll have to scroll to see them:
- Split "timeout" for reports into two: one for rendering, one for sending. Each of these operations could conceivably time out.
- Pass a `Duration timeout` into both `Renderer.render` and `Sender.send`.
- Inject a `ScheduledExecutorService` into Renderer/Sender implementations, in order to respect the passed timeout.

Tested:
- added automated tests to ensure timeouts are respected
- manually created and edited reports in the UI
- created a Grafana report pointed at a normal web page (which will never fire the appropriate event, and therefore time out); used PGAdmin to confirm that the report execution aborted+failed after the specified timeout

### CODE REVIEW

File: app/models/ebean/Report.java [hunk 1]
```diff
@@ -92,8 +92,11 @@
     @JoinColumn(name = "report_schedule_id")
     private ReportSchedule schedule;
 
-    @Column(name = "timeout_nanos")
-    private long timeoutNanos;
+    @Column(name = "render_timeout_nanos")
```
Comment (cwbriones) [inline app/models/ebean/Report.java]: Did you create a previous migration for this? Wouldn't this create duplicate columns?

File: app/com/arpnetworking/metrics/portal/reports/impl/chrome/grafana/BaseGrafanaScreenshotRenderer.java [hunk 2]
```diff
@@ -111,10 +110,13 @@ public URI getUri(final GrafanaReportPanelReportSource source) {
      * <ul>
      *   <li>{@code chromePath} -- the path to the Chrome binary to use to render pages.</li>
      * </ul>
+     * @param timeoutExecutor used to schedule timeouts on individual send operations
      */
-    @Inject
-    /* package private */ BaseGrafanaScreenshotRenderer(@Assisted final Config config) {
-        super(config);
+    /* package private */ BaseGrafanaScreenshotRenderer(
+            final Config config,
```
[the remark this answers is not in the document]
Comment (speezepearson) [reply app/com/arpnetworking/metrics/portal/reports/impl/chrome/grafana/BaseGrafanaScreenshotRenderer.java]: Exactly! Guice should never be trying to instantiate a `BaseGrafanaScreenshotRenderer`, only a subclass thereof.

File: test/java/com/arpnetworking/metrics/portal/reports/ReportExecutionContextTest.java [hunk 3]
```diff
@@ -153,11 +155,13 @@ public void testExecute() throws Exception {
         context.execute(EXAMPLE_REPORT, T1).toCompletableFuture().get();
         Mockito.verify(_emailSender).send(
                 ALICE,
-                ImmutableMap.of(HTML, mockRendered(EXAMPLE_REPORT, HTML, TIME_RANGE))
+                ImmutableMap.of(HTML, mockRendered(EXAMPLE_REPORT, HTML, TIME_RANGE)),
+                Duration.ofSeconds(10)
         );
         Mockito.verify(_emailSender).send(
```
[the remark this answers is not in the document]
Comment (speezepearson) [reply test/java/com/arpnetworking/metrics/portal/reports/ReportExecutionContextTest.java]: I don't feel a strong need to verify those! If the MockRenderers weren't invoked, then the MockSender wouldn't be invoked.

File: app/models/internal/impl/DefaultReport.java [hunk 4]
```diff
@@ -119,11 +131,21 @@ public String toString() {
...
     private final ImmutableSetMultimap<ReportFormat, Recipient> _recipients;
 
+    /**
+     * {@link #execute}ing a report involves: rendering it; sending it; and a little bit of miscellaneous work.
+     * The render- and send-timeouts are configurable on a per-report basis.
+     * The miscellaneous work is trivial, but we still need to account for it in {@link #getTimeout()},
+     *   or else {@link #execute} might get cancelled before we've given both those stages a fair shot.
+     * The non-rendering, non-sending work done by {@link #execute} should bevirtually guaranteed to finish in less than this much time.
+     */
+    private static final Duration TIMEOUT_SLOP = Duration.ofSeconds(5);
```
[the remark this answers is not in the document]
Comment (speezepearson) [reply app/models/internal/impl/DefaultReport.java]: Hrrm. I can see where you're coming from. How would you feel about outsourcing `Duration DefaultReport#getTimeout()` to `Duration ReportExecutionContext#getTimeout(Report)`?

**Pull request closed without merging**